File: sdks/csharp/tests/Sudomimus.Token.Tests/TestHelpers.cs
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using Sudomimus.Token;

namespace Sudomimus.Token.Tests;

/// <summary>
/// Mints compact RS256 fixtures matching the 4.0.0 token contract.
/// </summary>
internal static class TestHelpers
{
    public sealed record RsaKeyPair(string PublicKeyPem, string PrivateKeyPem);

    public static RsaKeyPair GenerateRsaKeyPair(int keySize = 2048)
    {
        using var rsa = RSA.Create(keySize);
        var publicPem = rsa.ExportSubjectPublicKeyInfoPem();
        var privatePem = rsa.ExportPkcs8PrivateKeyPem();
        return new RsaKeyPair(publicPem, privatePem);
    }

    public static string MintToken<THeader, TBody>(THeader header, TBody body, string privateKeyPem)
    {
        // @sudoo/jwt 3.6+ emits all three JWT segments as base64url, no padding.
        var headerSeg = ToBase64Url(JsonSerializer.SerializeToUtf8Bytes(header));
        var bodySeg = ToBase64Url(JsonSerializer.SerializeToUtf8Bytes(body));

        using var rsa = RSA.Create();
        rsa.ImportFromPem(privateKeyPem);
        var signingInput = Encoding.UTF8.GetBytes($"{headerSeg}.{bodySeg}");
        var signature = rsa.SignData(signingInput, HashAlgorithmName.SHA256, RSASignaturePadding.Pkcs1);
        var sigSeg = ToBase64Url(signature);

        return $"{headerSeg}.{bodySeg}.{sigSeg}";
    }

    private static string ToBase64Url(byte[] bytes)
    {
        return StripPadding(Convert.ToBase64String(bytes)).Replace('+', '-').Replace('/', '_');
    }

    /// <summary>base64url-encode a raw (already-serialized) segment string.</summary>
    public static string EncodeSegment(string raw) => ToBase64Url(Encoding.UTF8.GetBytes(raw));

    /// <summary>
    /// Build a structurally valid (3-segment) JWT from raw, pre-base64url
    /// header/body/signature payloads. The signature is not real — useful for
    /// exercising parse paths that run before signature verification.
    /// </summary>
    public static string MintRaw(string headerJson, string bodyJson, string signature = "sig")
        => $"{EncodeSegment(headerJson)}.{EncodeSegment(bodyJson)}.{EncodeSegment(signature)}";

    public static string MintAccessToken(string privateKeyPem, string applicationAnchor = "anchor-1")
    {
        var iat = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
        var header = new
        {
            alg = "RS256",
            typ = TokenVerifier.AccessTokenType,
            kid = "key-1",
        };
        var body = new
        {
            iss = "https://connect-api.sudomimus.com",
            aud = applicationAnchor,
            sub = "subject-1",
            sid = "session-1",
            jti = "access-1",
            iat,
            exp = iat + 3600,
        };
        return MintToken(header, body, privateKeyPem);
    }

    public static string MintRefreshToken(string privateKeyPem, string applicationAnchor = "anchor-1")
    {
        var iat = DateTimeOffset.UtcNow.ToUnixTimeSeconds();
        var header = new
        {
            alg = "RS256",
            typ = TokenVerifier.RefreshTokenType,
            kid = "key-1",
        };
        var body = new
        {
            iss = "https://connect-api.sudomimus.com",
            aud = applicationAnchor,
            sid = "session-1",
            jti = "refresh-1",
            iat,
            exp = iat + 30 * 24 * 3600,
            rotationVersion = 1,
        };
        return MintToken(header, body, privateKeyPem);
    }

    private static string StripPadding(string b64) => b64.TrimEnd('=');
}

File: sdks/java/token/src/test/java/com/sudomimus/token/TestHelpers.java
package com.sudomimus.token;

import com.fasterxml.jackson.databind.ObjectMapper;

import java.nio.charset.StandardCharsets;
import java.security.KeyPair;
import java.security.KeyPairGenerator;
import java.security.Signature;
import java.security.interfaces.RSAPrivateKey;
import java.security.interfaces.RSAPublicKey;
import java.time.Instant;
import java.util.Base64;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Mints compact RS256 fixtures matching the 4.0.0 token contract.
 */
final class TestHelpers {

    static final ObjectMapper MAPPER = new ObjectMapper();
    static final Base64.Encoder B64URL = Base64.getUrlEncoder().withoutPadding();

    private TestHelpers() {}

    static final class RsaKeyPair {
        final String publicPem;
        final RSAPrivateKey privateKey;
        final RSAPublicKey publicKey;

        RsaKeyPair(String publicPem, RSAPrivateKey privateKey, RSAPublicKey publicKey) {
            this.publicPem = publicPem;
            this.privateKey = privateKey;
            this.publicKey = publicKey;
        }
    }

    static RsaKeyPair generateRsaKeyPair() throws Exception {
        KeyPairGenerator gen = KeyPairGenerator.getInstance("RSA");
        gen.initialize(2048);
        KeyPair kp = gen.generateKeyPair();
        RSAPublicKey pub = (RSAPublicKey) kp.getPublic();
        RSAPrivateKey priv = (RSAPrivateKey) kp.getPrivate();
        String pem = "-----BEGIN PUBLIC KEY-----\n"
                + Base64.getMimeEncoder(64, "\n".getBytes(StandardCharsets.US_ASCII))
                        .encodeToString(pub.getEncoded())
                + "\n-----END PUBLIC KEY-----\n";
        return new RsaKeyPair(pem, priv, pub);
    }

    static String mintToken(Map<String, Object> header, Map<String, Object> body, RSAPrivateKey priv) throws Exception {
        String headerSeg = B64URL.encodeToString(MAPPER.writeValueAsBytes(header));
        String bodySeg = B64URL.encodeToString(MAPPER.writeValueAsBytes(body));
        byte[] signingInput = (headerSeg + "." + bodySeg).getBytes(StandardCharsets.UTF_8);

        Signature sig = Signature.getInstance("SHA256withRSA");
        sig.initSign(priv);
        sig.update(signingInput);
        String sigSeg = B64URL.encodeToString(sig.sign());

        return headerSeg + "." + bodySeg + "." + sigSeg;
    }

    static String mintAccessToken(RSAPrivateKey priv, String anchor) throws Exception {
        long iat = Instant.now().getEpochSecond();
        Map<String, Object> header = new LinkedHashMap<>();
        header.put("alg", "RS256");
        header.put("typ", TokenVerifier.ACCESS_TOKEN_TYPE);
        header.put("kid", "key-1");

        Map<String, Object> body = new LinkedHashMap<>();
        body.put("iss", "https://connect-api.sudomimus.com");
        body.put("aud", anchor);
        body.put("sub", "subject-1");
        body.put("sid", "session-1");
        body.put("jti", "access-1");
        body.put("iat", iat);
        body.put("exp", iat + 3600);

        return mintToken(header, body, priv);
    }

    static String mintRefreshToken(RSAPrivateKey priv, String anchor) throws Exception {
        long iat = Instant.now().getEpochSecond();
        Map<String, Object> header = new LinkedHashMap<>();
        header.put("alg", "RS256");
        header.put("typ", TokenVerifier.REFRESH_TOKEN_TYPE);
        header.put("kid", "key-1");

        Map<String, Object> body = new LinkedHashMap<>();
        body.put("iss", "https://connect-api.sudomimus.com");
        body.put("aud", anchor);
        body.put("sid", "session-1");
        body.put("jti", "refresh-1");
        body.put("iat", iat);
        body.put("exp", iat + 30L * 24 * 3600);
        body.put("rotationVersion", 1);

        return mintToken(header, body, priv);
    }
}
